src/RoslynPad.Avalonia/Folding/CShapRoslynFoldingStrategy.cs

```diff
@@ -0,0 +1,212 @@
+﻿using System.Text.RegularExpressions;
+using AvaloniaEdit.Folding;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+using TextDocument = AvaloniaEdit.Document.TextDocument;
+
+namespace RoslynPad.Folding;
+
+/// <summary>
+/// Folding strategy based on <see cref="Microsoft.CodeAnalysis.CSharp.FoldingStrategy"></see> Roslyn lexer
+/// </summary>
+public partial class CShapRoslynFoldingStrategy : FoldingStrategy
+{
+    readonly List<NewFolding> _foldings = [];
+    readonly string _defaultFoldingText = "...";
+    bool _startFolded;
+
+    /// <summary>
+    /// Folding SyntaxKind types <see cref="Microsoft.CodeAnalysis.CSharp.SyntaxKind"></see>
+    /// </summary>
+    readonly Dictionary<SyntaxKind, SyntaxKind> _pairTypes = new()
```

Comment:
Instead of reimplementing this functionality, you can create an interface for the Roslyn service `BlockStructureService` inside the RoslynPad.Roslyn project. That project has access to Roslyn internals, so you could expose it like I do for many other internal Roslyn services. Then just call `document.GetLanguageService<IBlockStructureService>(). GetBlockStructureAsync()` to get all the blocks.

Reply:
Fixed